### ISSUE TITLE
Files not syncing just because of non-allowed characters in the filename

### ISSUE THREAD

**Issue by jancborchardt** [owncloud/client#3406]
I get this so often – files can’t be synced because there’s a `:` or a `|` or someting like that in the filename.

Screenshots automatically named with the date and time, saved websites with the pipe bar in the title etc – I always need to rename that stuff and it’s really annoying. Also, I have no idea that the files are not synced because I don’t get notified. It’s hidden away in the Activity tab of the client dialog, which I rarely open (especially if I would not work on this project).

Now @dragotin @danimo @guruz @ogoffart I know you do not like automatically renaming files, but not syncing stuff I expect to be synced is not acceptable. What could be a solution?

**Comment by moscicki** [comment]
This is nothing new. It has been an issue reported for last 2 years or so but as far as I know this is being addressed currently. Renaming on the server is not an option -- it would lead to total mess and keep in mind the limitations of the fileid approach: https://github.com/cernbox/smashbox/blob/master/protocol/protocol.md#lifecycle-and-semantics-of-fileid-vs-path-vs-etag

**Comment by jancborchardt** [comment]
I know it's nothing new, and there are problems with renaming. However we need to find a solution here.

**Comment by danimo** [comment]
@jancborchardt: With ownCloud 8.1, the server can properly reject files _before_ they are being uploaded in case the backend cannot accept any of the meta data (e.g. file name). We are hence lo longer in need of a hard-coded black list for all OSes. Also. 8.1 is giving up its black list.

All that needs to be done is to make our client blacklist depend on wether or not we sync against 8.1, and only use it (for the local side) on Windows otherwise.

https://github.com/owncloud/client/pull/2885 does that, sans the 8.1 detection, which needs to be added before we can merge.

Right @DeepDiver1975?

**Comment by jancborchardt** [comment]
@danimo this however still results in the original problem – the file not being synced – right? Or what does the explicit rejection change?

**Comment by danimo** [comment]
@jancborchardt This is not a problem we can solve with a sledge-hammer approach. This is essentially a first step to allow syncing where it's possible and avoids renaming hacks in almost all situations.

The platform where this does not help is Windows. But it's a first step. For Windows, we would indeed need to do some mapping, but we first need local hashing an other measures to ensure a reliable mapping.

Also, Mac OS does still need some handling, because you can't have `AAA.txt` and `aaa.txt` at the same time, and because the length of a file (not the total path as on Windows) is limited to 254 bytes.

Note I am not opposing the need for an overall solution, I am just trying to split this into issues, so we can identify the work needed, and possible duplicate tickets.

**Comment by moscicki** [comment]
@danimo: how do you know that server rejects the upload because of unsupported characters in the file name? Is this documented somewhere?

I think the real problem is if there is a legitimate file that is not uploaded to the server (e.g. currently a file containing ":" character on linux or mac). This may lead to data loss because the file never ends up on the server. I think this is what is being addressed now.

The opposite (a file containing ":" on the server that is not downloaded to the Windows client) is not a critical problem but rather an inconvenience. For me leaving it up to the user to solve is an acceptable option in this case. Of course if the user is properly notified that some files are not downloaded and why.

**Comment by danimo** [comment]
@moscicki 
> how do you know that server rejects the upload because of unsupported characters in the file name?

Because it will tell us once we attempt an upload, right after the header has been transmitted.

> Is this documented somewhere?

@DeepDiver1975 drove this change, he would know.

> I think the real problem is if there is a legitimate file that is not uploaded to the server (e.g. currently a file containing ":" character on linux or mac).

As i said, this will be addressed with 8.1 on the server and a client that is aware if that (probably 2.0).

**Comment by moscicki** [comment]
@DeepDiver1975: what is the http response code and headers in this case, please?

**Comment by jancborchardt** [comment]
> I think the real problem is if there is a legitimate file that is not uploaded to the server (e.g. currently a file containing ":" character on linux or mac). This may lead to data loss because the file never ends up on the server. I think this is what is being addressed now.

That’s exactly my point @moscicki – we only support Linux as server and I’m on a Linux desktop as well (ok, GNU/Linux ;D). So there’s nothing stopping the sync to the server.

If Windows clients can handle it, I would say that’s a separate problem. But that should not prevent the legitimate file from being synced, right?

**Comment by danimo** [comment]
@jancborchardt Again, this problem is on its way out of the door. The interesting problems is handling the remaining problems.

**Comment by jancborchardt** [comment]
Ah ok, I didn’t understand. Thanks! :)

**Comment by ogoffart** [comment]
Done in 91525a7d33c531c04772376e833ed0238a8ee510
